lib/Sema/TypeCheckStmt.cpp

```diff
@@ -352,13 +355,13 @@ class StmtChecker : public StmtVisitor<StmtChecker, Stmt*> {
   };
 
   StmtChecker(TypeChecker &TC, AbstractFunctionDecl *AFD)
-    : TC(TC), TheFunc(AFD), DC(AFD), IsREPL(false) { }
+    : TC(TC), TheFunc(AFD), DC(AFD), IsREPL(false), IsBraceStmtFromTopLevelDecl(false) { }
 
   StmtChecker(TypeChecker &TC, ClosureExpr *TheClosure)
-    : TC(TC), TheFunc(TheClosure), DC(TheClosure), IsREPL(false) { }
+    : TC(TC), TheFunc(TheClosure), DC(TheClosure), IsREPL(false), IsBraceStmtFromTopLevelDecl(false) { }
 
   StmtChecker(TypeChecker &TC, DeclContext *DC)
-    : TC(TC), TheFunc(), DC(DC), IsREPL(false) {
+    : TC(TC), TheFunc(), DC(DC), IsREPL(false), IsBraceStmtFromTopLevelDecl(true) {
```

Comment:
This is not necessarily `true`.  You'll want to ask `isa<TopLevelCodeDecl>(DC)`.